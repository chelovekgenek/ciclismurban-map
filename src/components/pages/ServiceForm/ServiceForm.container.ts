import { connect } from "react-redux"
import { withFormik } from "formik"
import { withRouter } from "react-router-dom"

import {
  SelectedActions,
  getSelectedRoot,
  getServicesFetching,
  ServicesCreateActions,
  ServicesUpdateActions,
  SelectedGetActions,
} from "store/entities/locations"
import { TAppState } from "store/entities"
import { ExposeGroup, ServiceModel } from "models/location"
import { handleLocationFormSubmit, getHandlerLocationFormValidate } from "helpers"

import { IProps, ServiceForm } from "./ServiceForm"
import { ServiceFormValues } from "./ServiceForm.helper"

export type IStateProps = {
  services: {
    fetching: ReturnType<typeof getServicesFetching>
  }
  selected: ReturnType<typeof getSelectedRoot>
}

export interface IDispatchProps {
  create: typeof ServicesCreateActions.request
  update: typeof ServicesUpdateActions.request
  getSelected: typeof SelectedGetActions.request
  clearSelected: typeof SelectedActions.clear
}

const hocConnect = connect<IStateProps, IDispatchProps, IProps, TAppState>(
  state => ({
    services: {
      fetching: getServicesFetching(state),
    },
    selected: getSelectedRoot(state),
  }),
  {
    create: ServicesCreateActions.request,
    update: ServicesUpdateActions.request,
    getSelected: SelectedGetActions.request,
    clearSelected: SelectedActions.clear,
  },
)
const hocWithFormik = withFormik<IProps, ServiceFormValues>({
  enableReinitialize: true,
  mapPropsToValues: props => new ServiceFormValues(props.selected.data as ServiceModel),
  handleSubmit: handleLocationFormSubmit,
  validate: getHandlerLocationFormValidate(ServiceModel, [ExposeGroup.WRITE]),
  validateOnBlur: true,
})

export const ServiceFormContainer = withRouter(hocConnect(hocWithFormik(ServiceForm)))
