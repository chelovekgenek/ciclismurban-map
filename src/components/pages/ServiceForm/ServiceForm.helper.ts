import { MapOptions } from "helpers"
import { ServiceModel } from "models/location"

export class ServiceFormValues extends ServiceModel {
  constructor(props?: ServiceModel) {
    super()

    if (props) {
      this.title = props.title
      this.description = props.description
      this.point = props.point
      this.image = props.image
    }
  }

  title = ""
  description = ""
  point = MapOptions.position
  image = ""
}
