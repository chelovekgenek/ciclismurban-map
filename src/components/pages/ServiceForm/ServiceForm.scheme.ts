import { ServiceModel, WeeklyScheduleModel } from "@ciclismurban/models"
import { MapOptions } from "helpers"

export class ServiceFormValues extends ServiceModel {
  constructor(props?: ServiceModel) {
    super()

    if (props) {
      this.title = props.title
      this.description = props.description
      this.point = props.point
      this.image = props.image
      this.schedule = props.schedule
    }
  }

  title = ""
  description = ""
  point = MapOptions.position
  image = ""
  schedule: WeeklyScheduleModel = {
    mon: [],
    tue: [],
    wed: [],
    thu: [],
    fri: [],
    sat: [],
    sun: [],
  }
}
