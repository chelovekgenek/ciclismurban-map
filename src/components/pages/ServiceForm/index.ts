export { ServiceFormContainer as ServiceForm } from "./ServiceForm.container"
