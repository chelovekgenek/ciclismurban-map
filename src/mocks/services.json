[
  {
    "uuid": "a8ec2111-f270-47c4-aeee-3c735e26e689",
    "title": "ServiceA",
    "description": "Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem.",
    "image": "https://previews.123rf.com/images/kostymo/kostymo1707/kostymo170700115/82270238-logo-bike-service-a-bicycle-sprocket-and-hand-holding-a-wrench-on-the-sides-repair-tune-mountain-bik.jpg",
    "point": {
      "lat": 47.0265833,
      "lng": 28.8183946
    },
    "createdAt": "2019-07-31T23:17:36.069Z",
    "updatedAt": "2019-07-31T23:17:36.069Z"
  },
  {
    "uuid": "0da5a5d5-850c-4780-a0be-81e5494cb45e",
    "title": "ServiceB",
    "description": "Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?",
    "image": "https://www.sefiles.net/merchant/1809/images/site/service-center-hp-slimC1.jpg?t=1527182509653",
    "point": {
      "lat": 47.01933,
      "lng": 28.831819
    },
    "createdAt": "2019-07-31T23:18:23.005Z",
    "updatedAt": "2019-07-31T23:18:23.005Z"
  },
  {
    "uuid": "98a6bae0-1264-4fc2-8bad-7e47a111ea61",
    "title": "ServiceC",
    "description": "At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga.",
    "image": "https://cdn.shopify.com/s/files/1/0267/5681/products/Bicycle_Service-01-03-03.png?v=1527502897",
    "point": {
      "lat": 47.025933,
      "lng": 28.84
    },
    "createdAt": "2019-07-31T23:19:47.083Z",
    "updatedAt": "2019-07-31T23:19:47.083Z"
  }
]
