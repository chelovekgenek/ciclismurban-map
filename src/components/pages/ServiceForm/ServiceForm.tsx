import React, { useEffect, useMemo } from "react"
import { Form, FormikProps } from "formik"
import { RouteComponentProps } from "react-router-dom"
import { isEmpty } from "lodash-es"

import { App } from "components/generic/layout"
import { Input, Textarea, GoogleMap, Upload, WeeklySchedule } from "components/generic/form"
import { Button } from "components/generic/ui"

import { IStateProps, IDispatchProps } from "./ServiceForm.container"
import { ServiceFormValues } from "./ServiceForm.helper"
import { ServiceModel } from "models/location"

interface RouteParams {
  id: string
}
export interface IProps extends IStateProps, IDispatchProps, RouteComponentProps<RouteParams> {}

export const ServiceForm: React.FC<IProps & FormikProps<ServiceFormValues>> = ({
  submitForm,
  resetForm,
  isValid,
  touched,
  match,
  selected,
  getSelected,
  clearSelected,
  services,
}) => {
  useEffect(() => {
    if (match.params.id) {
      getSelected({ entity: "services", uuid: match.params.id })
    }
    return () => {
      clearSelected()
    }
  }, [match.params.id, getSelected, clearSelected])

  const actions = useMemo(() => {
    const reset = (
      <Button key="1" onClick={() => resetForm(selected.data as ServiceModel)} disabled={isEmpty(touched)}>
        Сбросить
      </Button>
    )
    const submit = (
      <Button key="2" type="primary" onClick={submitForm} disabled={!isValid} loading={services.fetching}>
        Сохранить
      </Button>
    )
    return selected.data ? [reset, submit] : [submit]
  }, [selected.data, services.fetching, isValid, submitForm, resetForm, touched])
  const title = useMemo(() => (selected.data ? "Редактировать сервис" : "Новый сервис"), [selected.data])
  return (
    <App
      content={{
        useHeader: true,
        useLayout: true,
        title,
        actions,
      }}
    >
      <Form>
        <Input label="Название" name="title" />
        <Textarea label="Описание" name="description" autosize={{ minRows: 1, maxRows: 4 }} />
        <WeeklySchedule label="Время работы" name="schedule" />
        <Upload label="Изображение" name="image" />
        <GoogleMap name="point" />
      </Form>
    </App>
  )
}
