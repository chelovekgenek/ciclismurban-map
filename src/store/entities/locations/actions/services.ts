import { action, payload } from "ts-action"

import { IState as IServicesState } from "../reducers/services"

export enum ServicesGetTypes {
  REQUEST = "LOCATIONS__SERVICES__GET__REQUEST",
  SUCCESS = "LOCATIONS__SERVICES__GET__SUCCESS",
  FAILURE = "LOCATIONS__SERVICES__GET__FAILURE",
}

export enum ServicesCreateTypes {
  REQUEST = "LOCATIONS__SERVICES__CREATE__REQUEST",
  SUCCESS = "LOCATIONS__SERVICES__CREATE__SUCCESS",
  FAILURE = "LOCATIONS__SERVICES__CREATE__FAILURE",
}

export enum ServicesUpdateTypes {
  REQUEST = "LOCATIONS__SERVICES__UPDATE__REQUEST",
  SUCCESS = "LOCATIONS__SERVICES__UPDATE__SUCCESS",
  FAILURE = "LOCATIONS__SERVICES__UPDATE__FAILURE",
}

export enum ServicesDeleteTypes {
  REQUEST = "LOCATIONS__SERVICES__DELETE__REQUEST",
  SUCCESS = "LOCATIONS__SERVICES__DELETE__SUCCESS",
  FAILURE = "LOCATIONS__SERVICES__DELETE__FAILURE",
}

type TState = Required<IServicesState>
type TStateData = TState["data"][number]

export const ServicesGetActions = {
  request: action(ServicesGetTypes.REQUEST),
  success: action(ServicesGetTypes.SUCCESS, payload<TState["data"]>()),
  failure: action(ServicesGetTypes.FAILURE, payload<TState["error"]>()),
}

export const ServicesCreateActions = {
  request: action(ServicesCreateTypes.REQUEST, payload<Partial<TStateData>>()),
  success: action(ServicesCreateTypes.SUCCESS, payload<TStateData>()),
  failure: action(ServicesCreateTypes.FAILURE, payload<TState["error"]>()),
}

interface TUpdatePayload {
  uuid: TStateData["uuid"]
  payload: Partial<TStateData>
}
export const ServicesUpdateActions = {
  request: action(ServicesUpdateTypes.REQUEST, payload<TUpdatePayload>()),
  success: action(ServicesUpdateTypes.SUCCESS, payload<TStateData>()),
  failure: action(ServicesUpdateTypes.FAILURE, payload<TState["error"]>()),
}

export const ServicesDeleteActions = {
  request: action(ServicesDeleteTypes.REQUEST, payload<TStateData["uuid"]>()),
  success: action(ServicesDeleteTypes.SUCCESS, payload<TStateData["uuid"]>()),
  failure: action(ServicesDeleteTypes.FAILURE, payload<TState["error"]>()),
}
