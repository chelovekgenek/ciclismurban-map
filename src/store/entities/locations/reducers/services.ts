import { reducer, on } from "ts-action"

import { ServiceModel } from "models/location"

import { ServicesGetActions, ServicesCreateActions, ServicesUpdateActions, ServicesDeleteActions } from "../actions"

export interface IState {
  data: ServiceModel[]
  fetching: boolean
  error?: Error
}

const initialState: IState = {
  data: [],
  fetching: false,
}

export default reducer(
  initialState,
  on(
    ServicesGetActions.request,
    ServicesCreateActions.request,
    ServicesUpdateActions.request,
    ServicesDeleteActions.request,
    (state: IState) => ({
      ...state,
      fetching: true,
    }),
  ),
  on(
    ServicesGetActions.failure,
    ServicesCreateActions.failure,
    ServicesUpdateActions.request,
    ServicesDeleteActions.failure,
    (state: IState, { payload }: any) => ({
      ...state,
      fetching: false,
      error: payload,
    }),
  ),
  on(ServicesGetActions.success, (state, { payload }) => ({ ...state, fetching: false, data: payload })),
  on(ServicesCreateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.concat(payload),
  })),
  on(ServicesUpdateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.map(item => (item.uuid === payload.uuid ? payload : item)),
  })),
  on(ServicesDeleteActions.success, (state, { payload }) => ({
    ...state,
    data: state.data.filter(item => item.uuid !== payload),
    fetching: false,
  })),
)
