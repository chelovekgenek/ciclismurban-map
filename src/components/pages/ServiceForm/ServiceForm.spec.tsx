import React from "react"
import { shallow } from "enzyme"

import formProps from "mocks/formik-form-props"
import services from "mocks/services.json"

import { ServiceFormValues } from "./ServiceForm.helper"
import { ServiceForm } from "./ServiceForm"

describe("pages/ServiceForm", () => {
  const initialValues = new ServiceFormValues()
  const render = (props: Partial<React.ComponentProps<typeof ServiceForm>> = {}) =>
    shallow(
      <ServiceForm
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        services={{ fetching: false }}
        selected={{ fetching: false }}
        getSelected={jest.fn() as any}
        clearSelected={jest.fn() as any}
        create={jest.fn() as any}
        update={jest.fn() as any}
        match={{ params: {} } as any}
        history={{} as any}
        location={{} as any}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
  it("should match snapshot if `selected.data` is provided", () => {
    expect(render({ selected: { data: services[0] } as any })).toMatchSnapshot()
  })
  it("should retreive dataset by `match.params.id`", () => {
    // TODO https://github.com/airbnb/enzyme/issues/2091
    // const props = {
    //   match: { params: { id: "id" } } as any,
    //   getSelected: jest.fn() as any,
    // }
    // render(props)
    // expect(props.getSelected).toHaveBeenCalledWith({ entity: "services" })
  })
  it("should be able to reset form if touched", () => {
    const wrapper = render({ selected: { data: services[0] } as any, touched: { title: true } })
    expect((wrapper.find("App").prop("content") as any).actions[0].props.disabled).toBe(false)
  })
})
