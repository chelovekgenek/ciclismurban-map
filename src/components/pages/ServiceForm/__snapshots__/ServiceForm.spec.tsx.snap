// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/ServiceForm should match snapshot 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Новый сервис",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <WeeklyScheduleWrapper
      label="Время работы"
      name="schedule"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;

exports[`pages/ServiceForm should match snapshot if \`selected.data\` is provided 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
        >
          Сбросить
        </Button>,
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Редактировать сервис",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <WeeklyScheduleWrapper
      label="Время работы"
      name="schedule"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;
